src/java/htsjdk/samtools/CRAMFileReader.java

```diff
@@ -567,47 +571,65 @@ public void close() {
 
         @Override
         public boolean hasNext() {
-            if (next != null) return true;
-            if (noMore) return false;
-
+            if (next != null) {
+                return true;
+            }
+            else if (!hasMore) {
+                return false;
+            }
             while (delegate.hasNext()) {
-                next = delegate.next();
-
-                if (isWithinTheInterval(next)) break;
-                if (isBeyondTheInterval(next)) {
-                    next = null;
-                    noMore = true;
-                    return false;
+                SAMRecord nextRecord = delegate.next();
+                if (matchesInterval(nextRecord)) {
+                    next = nextRecord;
+                    break;
+                }
+                else if (isBeyondTheInterval(nextRecord)) {
+                    hasMore = false;
+                    break;
                 }
-                next = null;
             }
-
             return next != null;
         }
 
-        boolean isWithinTheInterval(final SAMRecord record) {
-            final boolean refMatch = record.getReferenceIndex() == interval.referenceIndex;
-            if (interval.start == -1) return refMatch;
-            if (!refMatch) return false;
-
-            final int start = record.getAlignmentStart();
-            final int end = record.getAlignmentEnd();
-            if (overlap) {
-                return CoordMath.overlaps(start, end, interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end);
-            } else {
-                // contained:
-                return CoordMath.encloses(interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end, start, end);
+        private boolean matchesInterval(final SAMRecord samRecord) {
+            int refIndex = samRecord.getReferenceIndex();
+            if (refIndex == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX ||
+                    refIndex != interval.referenceIndex) {
+                return false;
+            }
+            else {
+                int alignStart = samRecord.getAlignmentStart();
+                int alignEnd = samRecord.getAlignmentEnd();
+                if (alignEnd == 0) {
+                    // handle placed but "unmapped" since they can pass the overlap test
+                    return (alignStart >= interval.start) &&
+                            (interval.end <= 0 || alignStart <= interval.end);
```

Comment:
Handle case where `alignStart < interval.start && alignEnd == 0` in a manner consistent with the BAM-querying code.